extensions/amp-analytics/0.1/events.js

```diff
@@ -409,6 +411,84 @@ export class IniLoadTracker extends EventTracker {
 
 
 /**
+ * Tracks timer events.
+ */
+export class TimerEventTracker extends EventTracker {
+  /**
+   * @param {!./analytics-root.AnalyticsRoot} root
+   */
+  constructor(root) {
+    super(root);
+    /** @const @private {!Array<number>} */
+    this.trackers_ = [];
+  }
+
+  /** @override */
+  dispose() {
+    const win = this.root.ampdoc.win;
+    this.trackers_.forEach(intervalId => {
+      win.clearInterval(intervalId);
+    });
+  }
+
+  /** @override */
+  add(context, eventType, config, listener) {
+    const timerSpec = config['timerSpec'];
+    user().assert(timerSpec && typeof timerSpec == 'object',
+        'Bad timer specification');
+    user().assert('interval' in timerSpec,
+        'Timer interval specification required');
+    const interval = Number(timerSpec['interval']) || 0;
+    user().assert(interval >= MIN_TIMER_INTERVAL_SECONDS,
+        'Bad timer interval specification');
+    const maxTimerLength = 'maxTimerLength' in timerSpec ?
+        Number(timerSpec['maxTimerLength']) : DEFAULT_MAX_TIMER_LENGTH_SECONDS;
+    user().assert(maxTimerLength == null || maxTimerLength > 0,
+        'Bad maxTimerLength specification');
+    const callImmediate = 'immediate' in timerSpec ?
+        Boolean(timerSpec['immediate']) : true;
+
+    const win = this.root.ampdoc.win;
+    const intervalId = win.setInterval(() => {
+      listener(this.createEvent_(eventType));
+    }, interval * 1000);
+    this.trackers_.push(intervalId);
+    win.setTimeout(() => {
+      this.removeTracker_(intervalId);
+    }, maxTimerLength * 1000);
+    if (callImmediate) {
+      listener(this.createEvent_(eventType));
+    }
+    return () => {
+      this.removeTracker_(intervalId);
+    };
+  }
+
+  /**
+   * @param {string} eventType
+   * @return {!AnalyticsEvent}
+   * @private
+   */
+  createEvent_(eventType) {
+    return new AnalyticsEvent(this.root.getRootElement(), eventType);
+  }
+
+  /**
+   * @param {number} intervalId
+   * @private
+   */
+  removeTracker_(intervalId) {
+    const win = this.root.ampdoc.win;
```

Comment:
worth to have a `this.win` param?

Reply:
Nay, it's inlined by compiler and we get a free resolution w/o biting the cost of an extra state variable.